wled00/FX.cpp

```diff
@@ -94,6 +95,50 @@ uint16_t mode_static(void) {
 static const char _data_FX_MODE_STATIC[] PROGMEM = "Solid";
 
 
+/*
+ * Copy selected segment and perform (optional) color adjustments
+ */
+// TODO: what is the correct place to put getRenderedPixelXY()?
+static uint32_t getRenderedPixelXY(Segment& seg, unsigned x, unsigned y = 0) {
+  // We read pixels back following mirror/reverse/transpose but ignoring grouping
+  // For every group-length pixels, add spacing
+  x *= seg.groupLength(); // expand to physical pixels
+  y *= seg.groupLength(); // expand to physical pixels
+  if (x >= seg.width() || y >= seg.height()) return 0;  // fill out of range pixels with black
+  #warning this check can be removed once 2D offset is fixed
+  uint32_t offset = seg.is2D() ? 0 : seg.offset; // dirty fix, offset in 2D segments should be zero but is not TODO: fix the offset
+  return strip.getPixelColorXY(seg.start + offset + x, seg.startY + y);
```

Comment:
This will get a pixel from another segment if you are overlapping segments and use spacing.

Reply:
The groupLength() call above should include the spacing?

Reply:
Correct. Needed to check sources.
Still it will not handle transposed segments correctly.

Reply:
It's supposed to copy the data post-transposition.  But I think you're correct that there's a bug here - if transpose is set, we need to swap virtualWidth() and virtualHeight() in the copy fx function.

Reply:
I tested transposed source segment (among many other things) it copies right (but maybe I missed a case)

Reply:
I think it might run in to trouble if the source segment x/y sizes aren't multiples of each other.  Eg, source = 3x8 transpose.

Reply:
So far I could not find a case where transpose does not work, tested various sizes (3x16, 5x16, 3x7) and its always correct. If segments are overlapping, the getRenderedPixelXY() will of course read whatever has been written to the buffer, so if a segment overwrites data, the copy function will still copy that. This can not be avoided and depends on order of segments (as does overlapping segments in general)